contracts/schemes/Competition.sol

```diff
@@ -0,0 +1,349 @@
+        address payable beneficiary = suggestions[_suggestionId].suggester;
+        if (msg.sender == suggestions[_suggestionId].suggester) {
+            //only suggester can redeem to other address
+            if (beneficiary != address(0)) {
+                beneficiary = _beneficiary;
+            }
+        }
+        _redeem(_suggestionId, beneficiary);
+    }
+
+    /**
+    * @dev setSnapshotBlock set the block for the reputaion snapshot
+    * @param _proposalId the proposal id
+    */
+    function setSnapshotBlock(bytes32 _proposalId) public {
+        // solhint-disable-next-line not-rely-on-time
+        require(proposals[_proposalId].votingStartTime < now, "voting period not started yet");
+        if (proposals[_proposalId].snapshotBlock == 0) {
+            proposals[_proposalId].snapshotBlock = block.number;
+            emit SnapshotBlock(_proposalId, block.number);
+        }
+    }
+
+    /**
+    * @dev getOrderedIndexOfSuggestion return the index of specific suggestion in the winners list.
+    * @param _proposalId proposal id
+    * @param _suggestionId suggestion id
+    */
+    function getOrderedIndexOfSuggestion(bytes32 _proposalId, uint256 _suggestionId)
+    public
+    view
+    returns(uint256 index) {
+        uint256[] memory topSuggestionsOrdered = proposals[_proposalId].topSuggestionsOrdered;
+        /** get how many elements are greater than a given element*/
+        for (uint256 i; i < topSuggestionsOrdered.length; i++) {
+            if (suggestions[topSuggestionsOrdered[i]].totalVotes > suggestions[_suggestionId].totalVotes) {
+                index++;
+            }
+        }
+    }
+
+    /**
+    * @dev refreshTopSuggestions this function maintain a winners list array.
+    * @param _proposalId proposal id
+    * @param _suggestionId suggestion id
+    */
+    function refreshTopSuggestions(bytes32 _proposalId, uint256 _suggestionId) private {
+        uint256[] storage topSuggestionsOrdered = proposals[_proposalId].topSuggestionsOrdered;
+        if (topSuggestionsOrdered.length < proposals[_proposalId].numberOfWinners) {
+            topSuggestionsOrdered.push(_suggestionId);
+        } else {
+         /** get the index of the smallest element **/
+            uint256 smallest = 0;
+            for (uint256 i; i < proposals[_proposalId].numberOfWinners; i++) {
+                if (suggestions[topSuggestionsOrdered[i]].totalVotes <
+                    suggestions[topSuggestionsOrdered[smallest]].totalVotes) {
+                    smallest = i;
+                }
+            }
+
+            if (suggestions[topSuggestionsOrdered[smallest]].totalVotes < suggestions[_suggestionId].totalVotes) {
+                topSuggestionsOrdered[smallest] = _suggestionId;
+            }
+        }
+    }
+
+    /**
+    * @dev redeem a winning suggestion reward
+    * @param _suggestionId suggestionId
+    * @param _beneficiary - the reward beneficiary
```

Comment:
honestly, the code in this fucntion is too complex for me to understand. But i do see it pays out to _beneficiary, but it does not seem to check any relation between the suggester of the suggestion and either msg.sender or _beneficiary.

Reply:
the beneficiary is the suggester .
(or otherwise if the suggester want it differently)

Reply:
will simplify this function